base/operators.jl

```diff
@@ -961,6 +961,7 @@ const => = Pair
 start(p::Pair) = 1
 done(p::Pair, i) = i>2
 next(p::Pair, i) = (getfield(p,i), i+1)
+eltype{A,B}(p::Pair{A,B}) = Union{A,B}
```

Comment:
Everything has an `eltype` (including things that are not iterable). For iterable objects, the `eltype` has to be a supertype of all elements of the iterable; otherwise functions like `collect` will not work. `Any` is a safe choice; `Union{A,B}` is a slightly better one, especially because this means that for `Pair{A,A}`, the (possibly-)concrete type `A` will be used instead of an abstract type.